main.py

```diff
@@ -0,0 +1,351 @@
+        dns_names = crypto_util.get_names_from_subject_and_extensions(csr.subject, csr.extensions)
+        dns_names = [name.lower() for name in dns_names]
+        try:
+            san_ext = csr.extensions.get_extension_for_class(x509.SubjectAlternativeName)
+        except x509.ExtensionNotFound:
+            ip_names = []
+        else:
+            ip_names = san_ext.value.get_values_for_type(x509.IPAddress)
+        return dns_names, ip_names
+
+    async def poll_authorizations(
+        self, orderr: messages.OrderResource, deadline: datetime.datetime
+    ) -> messages.OrderResource:
+        """Poll Order Resource for status.
+        This is copied from certbot's acme/client.py and modified to be async.
+        """
+        responses = []
+        for url in orderr.body.authorizations:
+            while datetime.datetime.now() < deadline:
+                authzr = self.client_acme._authzr_from_response(self.client_acme._post_as_get(url), uri=url)
+                if authzr.body.status != messages.STATUS_PENDING:  # pylint: disable=no-member
+                    responses.append(authzr)
+                    break
+                await asyncio.sleep(1)
+        # If we didn't get a response for every authorization, we fell through
+        # the bottom of the loop due to hitting the deadline.
+        if len(responses) < len(orderr.body.authorizations):
+            raise errors.TimeoutError()
+        failed = []
+        for authzr in responses:
+            if authzr.body.status != messages.STATUS_VALID:
+                for chall in authzr.body.challenges:
+                    if chall.error is not None:
+                        failed.append(authzr)
+        if failed:
+            raise errors.ValidationError(failed)
+        return orderr.update(authorizations=responses)
+
+    async def poll_finalization(
+        self,
+        orderr: messages.OrderResource,
+        deadline: datetime.datetime,
+        fetch_alternative_chains: bool = False,
+    ) -> messages.OrderResource:
+        """
+        Poll an order that has been finalized for its status.
+        If it becomes valid, obtain the certificate.
+        It taken from certbot's acme/client.py and modified to be async.
+
+        :returns: finalized order (with certificate)
+        :rtype: messages.OrderResource
+        """
+
+        while datetime.datetime.now() < deadline:
+            await asyncio.sleep(1)
+            response = self.client_acme._post_as_get(orderr.uri)
+            body = messages.Order.from_json(response.json())
+            if body.status == messages.STATUS_INVALID:
+                if body.error is not None:
+                    raise errors.IssuanceError(body.error)
+                raise errors.Error(
+                    "The certificate order failed. No further information was provided by the server."
+                )
+            elif body.status == messages.STATUS_VALID and body.certificate is not None:
```

Comment:
в if выше всегда по raise выйдет:
```suggestion
            if body.status == messages.STATUS_VALID and body.certificate is not None:
```